contracts/Lottery.sol

```diff
@@ -19,72 +20,103 @@ contract Lottery {
+        playerMap[msg.sender] = _player;
+        players.push(_player);
         
-        result = "FAIL !!!";
-        if(!playerMap[msg.sender].isExist ) {
-            Player memory _player = Player(msg.sender, _name, _emailId, _phoneNumber,true);
-            playerMap[msg.sender] = _player;
-            players.push(_player);
-            result = "Registered Successfully !!!";
-        }
-
-        return(result);
         // new event to be send to owner
+        emit NewUser(msg.sender);
+
     }
     
-    function startLottery(uint32 _ticketPrice) public{
+    function startLottery(uint _ticketPrice) public{
         require (owner == msg.sender, "Only owner can open lottery process.");
-        state = LotteryState.Started;
         ticketPrice = _ticketPrice;
+        state = LotteryState.Started;
     }
-
-    function stopLottery() public{
-        require (owner == msg.sender, "Only owner can open lottery process.");
-        state = LotteryState.Finished;
+    
+    function terminateLottery() public {
+        require (owner == msg.sender, "Only owner can close lottery.");
+        require (state == LotteryState.Started, "To cancel lottery contract, it should be in Started state");
+        state = LotteryState.Terminated;
+        emit LotteryStatusUpdateWithMessage(msg.sender, getLotteryState(), "Lottery has been terminated, you will receive refund shortly.");
+        
+        //Refund process for every player
+        for (uint counter = 0; counter < players.length; counter++) {
+            Player memory player = players[counter];
+            LotteryTicket[] memory tickets = playerLotteryTicketsMapping[player.playerAddress];
+            uint playerAmount = tickets.length * ticketPrice;
+            if(playerAmount > 0) {
+                player.playerAddress.transfer(playerAmount);
+                emit RefundLottery(player.playerAddress, playerAmount);
+            }
+        }
         
     }
     
-    function buyLotteryTickets(uint32 _ticketNumbers) public payable returns (string memory result) {
-        require(msg.value >= (_ticketNumbers*ticketPrice), "Insufficient balance.");
-        require(state == LotteryState.Started, "Lottery is not yet opened");
+    
+    function buyLotteryTickets(uint count, uint value) public payable {
         require(owner != msg.sender, "Lottery owner/organizer can't buy tickets.");
+        require(state == LotteryState.Started, "Lottery is not yet opened");
         require(playerMap[msg.sender].isExist,"User not registered.");
+        require(msg.sender.balance > value, "Insufficient balance to purchase ticket(s).");
+        //uint memory ticketsPrice = count.mul(ticketPrice);
+        uint  ticketsPrice = count * ticketPrice;
+        require(value == ticketsPrice, "There is difference between actual price and provided price.");
+        require(msg.value == ticketsPrice, "Amount sent for buying ticket(s) are not sufficient");
```

Comment:
These checks are a bit confusing

`value` parameter seems like the total amount needed to purchase the tickets. But if we already send it as part of the `msg` object, then do we really need it as a function parameter too?

Also, the checks on lines 99 and 103 seem to be similar (checking for sufficiency of balance) but they seem to have different conditions in place. Any specific reasons.

Reply:
Valid point, i had two different implementation in my mind. Reading from msg.value is 2nd option, making it consistent